cmd/log/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use humility::cli::Subcommand;
+use humility::core::Core;
+use humility::hubris::*;
+use humility::reflect::{self, Load, Value};
+use humility_cmd::doppel::{StaticCell, Stringbuf};
+use humility_cmd::{Archive, Attach, Command, Validate};
+use std::str;
+use std::thread;
+use std::time;
+
+#[derive(Parser, Debug)]
+#[clap(name = "stringbuf", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct StringbufArgs {
+    /// list all stringbufs
+    #[clap(long, short)]
+    list: bool,
+    /// continously update the log
+    #[clap(long, short, conflicts_with = "list")]
+    monitor: bool,
+    /// how long to delay in ms between polling for new log entries
+    #[clap(
+        long,
+        short,
+        default_value = "90",
+        requires = "monitor",
+        conflicts_with = "list"
+    )]
+    delay: u64,
+    /// print only a single stringbuffer by substring of name
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+fn stringbuf_monitor(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    delay: Option<u64>,
+    once: bool,
+) -> Result<()> {
+    let mut buf: Vec<u8> = vec![];
+    buf.resize_with(ringbuf_var.size, Default::default);
+
+    let mut update_stringbuf = || -> Stringbuf {
+        let _info = core.op_start().unwrap();
+        core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+        core.op_done().unwrap();
+
+        // Load the stringbuf struct with data from buf
+        let ringbuf_val: Value = Value::Struct(
+            reflect::load_struct(hubris, &buf, definition, 0).unwrap(),
+        );
+
+        let cell: StaticCell = StaticCell::from_value(&ringbuf_val).unwrap();
+        Stringbuf::from_value(&cell.cell.value).unwrap()
+    };
+
+    //
+    // TODO
+    // set ctrl c handler before looping
+    // this will ensure that op_done is always called even if interrupted
+    // see jira https://rivosinc.atlassian.net/browse/SW-440
+    //
+    // represents the last seen values from the string buffer
+    let mut last_last: u32 = 0;
+    let mut last_buffer = vec![];
+    loop {
+        //
+        // Read the log from the core and convert to a u8 slice
+        //
+        let stringbuf = update_stringbuf();
+        let buffer = stringbuf.buffer.as_slice();
+        let last = stringbuf.last.unwrap();
+
+        // don't print duplicates
+        if last_last == last && last_buffer == stringbuf.buffer {
+            println!("skipped");
+            continue;
+        }
+
+        // read everything between current last, and last last by
+        // generating index range which connects these two in a circular manner, only including the
+        // new characters, and in the order they should be printed
+        // last_last has been printed already, last has not
+        let new_idxs: Vec<u32> = if last > last_last {
```

Comment:
If the last index wraps around and goes past last_last, wouldn't you skip printing 0->last_last?

Reply:
if last > last_last, ex:
```
['t', 'a', 's', 'k', ' ', 'f', 'a', 'u', 'l', 't']
       ^                             ^
      last_last                     last
```
Then we need to print "sk fau", we do not need to print "lt" or "t".
I am not sure If I  follow your question though, or if it was fixed with the new looping logic.  Take a look and let me know if you still have questions/concerns.

Reply:
I was thinking about this case:

Time instance # 0 - Fetch ringbuffer
```
['t', 'a']
       ^          
      last                   
      last_last                    
```

i.e. you printed "ta".

Time instance # 1 - Haven't fetched ringbuffer yet, and the buffer overflows
```
['t', 'a', 's', 'k', ' ', 'f', 'a', 'u', 'l', 't', ' ', 'a', 'n', 'o', 't', 'h', 'e', 'r', 'f', 'a', 'u', 'l', 't']
       ^                                                                                                         ^
      last_last                                                                                                 last
```

Time instance # 2 (buffer has overflown) - Fetch ringbuffer
```
['n', 'e', 'w', ' ', 'f', 'a', 'u', 'l', 't', '.', ' ', 'a', 'n', 'o', 't', 'h', 'e', 'r', 'f', 'a', 'u', 'l', 't']
       ^                                  ^
      last_last                          last
```

Here, we will print "ew fault", whereas we should have printed: "another fault new fault"

I am thinking about the overflow case. Maybe we dont want to bother about it right now?

Reply:
Ah, I see.  I think we should be able to detect overflow by seeing if the character pointed to by last_last changed.  I created a [jira](https://rivosinc.atlassian.net/browse/SW-1056) to track this, I can add better support for overflows in another PR.

Reply:
Ack.